### PR TITLE
Fix for ES Module detection using npm@7 (#4295)

### PR DESCRIPTION
**Pull request by richardsimko**

### CODE REVIEW

File: lib/migrations/util/is-module-type.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    try {
+        if (process.env.npm_package_json) {
+            // npm >= 7.0.0
+            const packageJson = JSON.parse(await readFile(process.env.npm_package_json, 'utf-8'));
+            if (packageJson.type === 'module') {
+                return true;
+            }
+        }
+    } catch (e) {
+        console.warn('Error reading package.json', e);
```
Comment (kibertoad) [inline lib/migrations/util/is-module-type.js:13]: In which scenarios do you expect this logic to fail?

Comment (richardsimko) [reply lib/migrations/util/is-module-type.js:13]: Now that I think about it it should not be something that can occur, as `npm` would refuse to run if `package.json` had an invalid syntax.

**Comment by kibertoad** [comment]
@richardsimko Bunch of previous tests seem to be failing, see https://github.com/knex/knex/pull/4296/checks?check_run_id=1904682577

**Comment by richardsimko** [comment]
Yeah I don't think my approach to mocking the env vars was good. I'll try another way but if that doesn't work I'm not sure it's possible to add tests for this code in a good way.

**Comment by kibertoad** [comment]
@richardsimko Unit tests are fixed, but it looks like something in migration loading logic might be broken, judging by non-mysql integration tests.

**Comment by richardsimko** [comment]
I think what's most likely happening is that I've uncovered latent issues where `importFile` was not used with `await`. `import(filePath)` is an async function and always was - but this function was only used in the ESM case which doesn't seem to be commonly used. Now `importFile` is always async and this is how that issue manifests. Fixing this would require a lot more in-depth knowledge of knex than what I have but if you have any pointers as to where this might be an issue I'm all ears.

**Comment by richardsimko** [comment]
`importFile` seems to be used here https://github.com/richardsimko/knex/blob/2cc3247d83e17cfb7d5788f6b391ab31faa4758b/lib/migrations/migrate/sources/fs-migrations.js#L84:L84

Which in turn is used in many places without `await`:
https://github.com/knex/knex/blob/2cc3247d83e17cfb7d5788f6b391ab31faa4758b/lib/migrations/migrate/Migrator.js#L493:L493
https://github.com/knex/knex/blob/2cc3247d83e17cfb7d5788f6b391ab31faa4758b/lib/migrations/migrate/Migrator.js#L145:L145
https://github.com/knex/knex/blob/2cc3247d83e17cfb7d5788f6b391ab31faa4758b/lib/migrations/migrate/Migrator.js#L86:L86

**Comment by kibertoad** [comment]
@richardsimko Yeah, noticed same thing: https://github.com/knex/knex/blob/b20a31fe5d0256234e1542c1ec4cbe2082f9970e/lib/migrations/migrate/sources/fs-migrations.js#L84 seems to be the only place where we importFile without awaiting, so I would review all cases where `getMigration` is used to ensure we await properly.

**Comment by richardsimko** [comment]
I can slap `await`s on all the above mentioned locations but in my experience fixing these types of issues is never that easy 😅

**Comment by kibertoad** [comment]
@richardsimko You might need to make few more functions async (e. g. `waterfallBatch`), but that should do it. We have pretty good coverage for migrations at this point, if something breaks down, you'll know immediately.

File: lib/migrations/migrate/Migrator.js [hunk 2]
```diff
@@ -80,12 +80,16 @@ class Migrator {
...
+      !(
+        await Promise.all(
+          migrations.map(async (migration) => {
+            const migrationContents = await this.config.migrationSource.getMigration(
+              migration
+            );
+            return !this._useTransaction(migrationContents);
+          })
+        )
+      ).some((e) => e);
```
Comment (kibertoad) [inline lib/migrations/migrate/Migrator.js:92]: Maybe a slightly more descriptive name could be used? from my understanding the value you'll have there is `isTransactionUsed` or something along these lines

**Comment by richardsimko** [comment]
It worked 🎉  I added a note in the changelog as well. Do you want me to squash the commits?

**Comment by kibertoad** [comment]
@richardsimko Thanks a lot! 
Squashing is not worth the effort, I'll squash on merge anyway.

**Review by kibertoad — APPROVED** [review]

**Comment by richardsimko** [comment]
@kibertoad Is there anything I have to do in order to get this backported to 0.21.x?

**Comment by kibertoad** [comment]
@richardsimko There is a 0.21 branch, just open a PR targetting that.

**Comment by richardsimko** [comment]
Thanks, I opened #4308

**Pull request merged**